### ISSUE TITLE
Upload is assigning "pileup" as autodetected datatype in a greedy way for tabular format data

### ISSUE THREAD

**Issue by jennaj** [galaxyproject/galaxy#3001]
Trigger to autodetect pileup format needs to be more discriminatory. 

Examples of upload text given "pileup" format (when are not - just undefined tabular) are in this history. To retest, copy text and upload again as pasted data using option to autodetect format: https://test.galaxyproject.org/u/jenjackson/h/tabspace-issue-testing

Discovered while testing this other tabular data ticket: https://github.com/galaxyproject/galaxy/issues/2602

Shared test history: https://test.galaxyproject.org/u/jenjackson/h/tabspace-issue-testing

Workaround: Do not use autodetect and assign tabular at upload OR reassign datatype per dataset after upload.

@guerler is this something you could help with? Or is it @dannon 's ?

**Comment by nsoranzo** [comment]
The pileup format is very loosely defined, so the sniffer check only for a few characteristics:
- at least 3 columns (this may be increased to 5, I think)
- second column is an integer (may also check that it is `>= 0`)
- third column is one of 'A', 'C', 'G', 'T', 'N', 'a', 'c', 'g', 't', 'n'

Apart from the changes proposed above, I'm not sure this is a bug in the sniffer.

**Comment by jennaj** [comment]
Agree, maybe this is enough

> at least 3 columns (this may be increased to 5, I think)

**Comment by jennaj** [comment]
All - when on Main, please ping me and I will test and close out. Thanks so MUCH

@jmchilton @natefoo @martenson

**Comment by jennaj** [comment]
This is still a problem at http://usegalaxy.org. See https://usegalaxy.org/u/jen/h/test-greedy-pileup-auto-detect-format-with-upload

@nsoranzo I wonder if your change never made it to Main?

**Comment by nsoranzo** [comment]
@jennaj Pretty sure the changes from #3033 have been on Main for some time. Any chance you can make that history public?

**Comment by jennaj** [comment]
@nsoranzo Yes, I thought so too then did another direct test and this came out. I don't know how that link got unshared but I just changed it back to share. 

Reran two more tests in that history, the same result. Details in history comments.

**Comment by nsoranzo** [comment]
@jennaj Trying to access https://usegalaxy.org/u/jen/h/test-greedy-pileup-auto-detect-format-with-upload still gives me "History is not accessible by user".

**Comment by jennaj** [comment]
@nsoranzo I check and it flipped back to unshared! Might be another issue. So, I just made both the link accessible and published on Main. Please try now. Once you see, I'll want to unpublish it (but can leave shared, if it will stick!)

**Comment by jennaj** [comment]
@davebx @guerler Do we need to update Main for this? I retested again and had pileup assigned as a datatype. Same test history as above. Galaxy `17.09` (pre-release at https://usegalaxy.org) also here: https://usegalaxy.org/u/jen/h/test-greedy-pileup-auto-detect-format-with-upload

Dataset 5 & 8 are exactly the same - 5 used autodetect (and was assigned pileup), 8 I manually assigned to tabular (and it stuck).

or maybe I am misunderstanding @nsoranzo ? 

_Please note_ that this came up for set of irregular datasets (users do load these) - three test tabular dataset with an inconsistent number of columns per line (each test was formatted slightly different to test another function, but seemed good test cases for this as well). That formatting might be confusing autodetect. But tabular would still be a better default assignment than pileup.

ps: The "shared then the spontaneously-unshared" issue came up for me again just this week. I asked @natefoo about it. There isn't a ticket for it but will try to capture more "what exactly I was doing" details if comes up again and then consider a ticket. Doesn't happen often.

**Comment by jennaj** [comment]
Just retested and it is assigning tabular now. 

Compare datasets 3-4-5 run earlier today with dataset 8 run just a bit after. Same exact input was used for datasets 5 & 8. Did something get fixed today? Don't see the pull request, but know there were some merged for 17.09. 

I am going to let it sit overnight then try again tomorrow - maybe 10x, same input, and see what happens.

**Comment by jmchilton** [comment]
I can see the file - it matches the description of a Pileup @nsoranzo described above - it has at least 5 columns, the second is an integer, the third is a base. @jennaj I think you need to specify a criteria the file doesn't have that we should look for. Should we verify that all the rows have the same number of non-empty column values? Should we write a heuristic to try to guess if the first column is a chromosome identifier? Should we disable sniffing for this datatype - is not something people upload?

**Comment by jennaj** [comment]
@jmchilton sorry for the delay - I like this idea. Skip sniffing for pileup, assign tabular instead. Tabular is still correct, just not as specific. People can assign pileup if they really need to (in Upload or Edit Attr). In all the histories I've looked at the pileup data was generated by macs within a Galaxy history, or copied from a Galaxy history/library - NOT uploaded.
